src/blob.jl

```diff
@@ -63,36 +63,54 @@ The number of bytes needed to allocate `T` itself.
 
 Defaults to `sizeof(T)`.
 """
-@generated function self_size(::Type{T}) where T
-    @assert isconcretetype(T)
+Base.@assume_effects :foldable function self_size(::Type{T}) where T
+    # This function is marked :foldable to encourage constant folding for this types-only
+    # static computation.
     if isempty(fieldnames(T))
-        quote
-            $(Expr(:meta, :inline))
-            $(sizeof(T))
-        end
+        sizeof(T)
     else
-        quote
-            $(Expr(:meta, :inline))
-            $(+(0, @splice i in 1:length(fieldnames(T)) begin
-                self_size(fieldtype(T, i))
-            end))
+        # Recursion is the fastest way to compile this, confirmed with benchmarks.
+        # Alternatives considered:
+        # - +(Iterators.map(self_size, fieldtypes(T))...)
+        # - runtime_size for-loop (below).
+        # Splatting is always slower, and breaks after ~30 fields.
+        # The for-loop is faster after around 15-30 fields, so we pick an
+        # arbitrary cutoff of 20:
+        if fieldcount(T) > 20
+            runtime_size(T)
+        else
+            _sum_field_sizes(T)
         end
     end
 end
+runtime_size(::Type{T}) where T =
+    let out = 0
+    for f in fieldtypes(T)
+        out += Blobs.self_size(f)
+    end
+    out
+end
```

Comment:
Perhaps the following would be easier for most folks to parse (or is there something special about the existing code that I'm missing?)
```suggestion
function runtime_size(::Type{T}) where T
    out = 0
    for f in fieldtypes(T)
        out += Blobs.self_size(f)
    end
    out
end
```

Reply:
thanks, that's better. I wrote the function at the REPL and i was being lazy, then i forgot. :) 👍